core/languages.lua

```diff
@@ -9,30 +13,124 @@ end)
 
 local loadonce = {}
 
+-- BEGIN OMIKHLEIA HACKLANG
+local icu = require("justenoughicu")
+
+-- This small utility could be moved to utilities as SU.forLanguage()...
+-- The input is expected to be a valid BCP47 canonical language.
+-- The idea is to find the "closest" language accepted by the callback:
+-- Loop removing a language specifier until the callback returns a non-nil value
+-- Returns that value and the matched language in that case, or nil no callback matched.
+-- E.g. "xx-Xxxx-XX" will be matched against "xx-Xxxx--XX", "xx-Xxxx", "xx" until one of
+-- these are satisfied.
+-- Returns
+--    nil if not callback could process the language
+--    language, res if a callback could (returning the matched language pattern and the
+--    result of the callback)
+local function forLanguage(langbcp47, callback)
+  while langbcp47 do
+    local res = callback(langbcp47)
+    if res then
+      return res, langbcp47
+    end
+    langbcp47 = langbcp47:match("^(.+)-.*$") -- split at dash (-) and remove last part.
```

Comment:
I think this matching/parsing bit needs to be in cldr-lua.

Reply:
Wherever it eventually ends, the utility is also useful for other packages. E.g. I wish I had it for my "smartquotes" package.

Reply:
The ICU bits should probably get moved from the font code to some base language support module, and yes should be setup in a way that every package can use them. This bit which isn't really ICU still looks like it goes in CLDR to me and could also return information about what the segments *mean*.